File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

    static class Edge {
        int start;
        int end;
        int weight;

        public Edge(int start, int end, int weight) {
            this.start = start;
            this.end = end;
            this.weight = weight;
        }
    }

    private static final int INF = 987654321;

    static int n;
    static List<List<Edge>> graph = new ArrayList<>();
    static int[] dist, history;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        int m = Integer.parseInt(br.readLine());
        StringTokenizer st;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            graph.get(from).add(new Edge(from, to, weight));
        }
        st = new StringTokenizer(br.readLine());
        int source = Integer.parseInt(st.nextToken());
        int destination = Integer.parseInt(st.nextToken());
        dijkstra(source, destination);
        StringBuilder sb = new StringBuilder();
        sb.append(dist[destination]).append('\n');

        Stack<Integer> stack = new Stack<>();
        int temp = destination;
        while (temp != -1) {
            stack.push(temp);
            temp = history[temp];
        }
        sb.append(stack.size()).append('\n');

        while (!stack.isEmpty()) {
            sb.append(stack.pop()).append(' ');
        }
        System.out.println(sb);
    }

    private static void dijkstra(int source, int destination) {
        dist = new int[n + 1];
        history = new int[n + 1];
        visited = new boolean[n + 1];
        Arrays.fill(dist, INF);
        dist[source] = 0;

        Queue<Edge> queue = new PriorityQueue<>(new Comparator<Edge>() {
            @Override
            public int compare(Edge o1, Edge o2) {
                return o1.weight - o2.weight;
            }
        });
        queue.offer(new Edge(-1, source, 0));

        while (!queue.isEmpty()) {
            Edge cur;
            do {
                cur = queue.poll();
            } while (!queue.isEmpty() && visited[cur.end]);
            if (visited[cur.end]) break;
            visited[cur.end] = true;
            history[cur.end] = cur.start;

            for (Edge edge : graph.get(cur.end)) {
                if (dist[edge.end] > dist[cur.end] + edge.weight) {
                    dist[edge.end] = dist[cur.end] + edge.weight;
                    queue.offer(new Edge(cur.end, edge.end, dist[edge.end]));
                }
            }
        }
    }

}

File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.cc
#include <iostream>
#include <limits.h>
#include <queue>
#include <functional>
#include <memory.h>
#include <stack>

using namespace std;

const int INF = INT_MAX;
int n, m;
vector< vector< pair<int, int> > > graph;

struct cmp {
    bool operator()(pair<int, int> a, pair<int, int> b) {
        return a.first > b.first;
    }
};

void dijkstra(int node, int to) {
    int dist[n + 1];
    for (int i = 0; i <= n; ++i) dist[i] = INF;
    dist[node] = 0;

    int path[n + 1];
    memset(path, -1, sizeof(path));
    path[node] = node;

    priority_queue< pair<int, int>, vector< pair<int, int> >, cmp > pq;
    pq.push(make_pair(0, node));

    bool visited[n + 1];
    memset(visited, 0, sizeof(visited));
    while (!pq.empty()) {
        pair<int, int> cur = pq.top();
        pq.pop();
        int d = cur.first, now = cur.second;
        if (visited[now]) continue;
        visited[now] = true;

        for (auto &edge : graph[now]) {
            int next = edge.first, nextD = edge.second;
            if (dist[next] > nextD + d) {
                dist[next] = nextD + d;
                pq.push(make_pair(dist[next], next));
                path[next] = now;
            }
        }
    }

    cout << dist[to] << '\n';

    stack<int> stack;
    while (to != node) {
        stack.push(to);
        to = path[to];
    }
    stack.push(node);
    cout << stack.size() << '\n';
    while (!stack.empty()) {
        cout << stack.top() << ' ';
        stack.pop();
    }
}

int main() {
    cin >> n >> m;

    int from, to, cost;
    graph = vector< vector< pair<int, int> > >(n + 1);
    for (int i = 0; i < m; ++i) {
        cin >> from >> to >> cost;
        graph[from].emplace_back(to, cost);
    }

    int start, end;
    cin >> start >> end;

    dijkstra(start, end);

    return 0;
}